KAFKA-6254: Incremental fetch requests

cmccabe opened the pull request.

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -533,41 +535,38 @@ class KafkaApis(val requestChannel: RequestChannel,
           response
         }
 
+        logger.info("WATERMELON: about to send Fetch response with parts.size={}, metadata={}",
```
Is this temporary?

cmccabe replied:
Fixed

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,645 @@
+class SessionlessFetchContext(val fetchData: util.Map[TopicPartition, FetchRequest.PartitionData]) extends FetchContext {
+  override def getFetchOffset(part: TopicPartition) = Option(fetchData.get(part)).map(_.fetchOffset)
+
+  override def foreachPartition(fun: (TopicPartition, FetchRequest.PartitionData) => Unit) = {
```
Should the return type of the method be Unit?

junrao replied:
Has the return type been changed to Unit yet?

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,645 @@
+        partMap.add(newCachedPart)
+        numAdded = numAdded + 1
+      } else {
+        if (reqData.maxBytes == 0) {
```
It's a bit unintuitive to use 0 maxBytes as an indication for removal.

cmccabe replied:
This seems similar to setting everything to 0 when there is a partition error, right?  It would be clearer if our RPC type system supported a more advanced type system.

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,645 @@
+  }
+
+  // Update the cached partition data based on the request.
+  def update(fetchData: FetchSession.REQ_MAP, reqMetadata: JFetchMetadata) = synchronized {
```
Since all the callers are already synchronizing on the session object, do we need to synchronize here?

cmccabe replied:
It's not technically needed, but it makes the code much clearer because the locking is consistent.  It also should have a very small overhead

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,645 @@
+                      var fetchOffset: Long,
+                      var lastStableOffset: Long,
+                      var fetcherLogStartOffset: Long,
+                      var localLogStartOffset: Long,
```
For consistency, perhaps it's better to either add local to all offsets or leave it out for all.

cmccabe replied:
The main reason it's on this one is to distinguish from fetcherLogStartOffset (the LSO of the follower, which is different from ours).  Maybe I should add "local" to all of them, though?

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,645 @@
+      mustRespond = true
+      localLogStartOffset = respData.logStartOffset
+    }
+    if (!Objects.equals(abortedTransactions, respData.abortedTransactions)) {
```
This is probably not needed. We only return abortedTransactions if we return non-empty data.

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,645 @@
+      // Partitions with new data are always included in the response.
+      mustRespond = true
+    }
+    if (lastStableOffset != respData.lastStableOffset) {
```
This is probably not needed. lastStableOffset is only relevant in EOS mode. In that mode, just checking if there is new data is enough. We do need to track HW since it can change even if there is no data change (for the follower). Propagating the latest HW to follower is useful for metrics such as MaxLag.

cmccabe replied:
Good point.  As per our discussion, I removed `lastStableOffset` and `abortedTransactions`, and added `highWatermark`.

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java:
```diff
@@ -0,0 +1,348 @@
+                return slot;
+            }
+            if (element.equals(newElement)) {
+                return INVALID_INDEX;
```
The callers of add() don't check the return value. Would it be better to just assert here?

cmccabe replied:
Hmm.  It's probably good to keep the existing add function as-is, beacuse then we implement the familiar Java contract for `java.util.Set`.  However, let me add a new method called `mustAdd` which throws an exception if the addition could not be performed, so that we catch any bugs here.

junrao commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java:
```diff
@@ -0,0 +1,348 @@
+    }
+
+    @Override
+    public boolean remove(Object o) {
```
Hmm, why would o be Object instead of E?

cmccabe replied:
Unfortunately, `java.util.Set` was added to the Java language before generics were added to the Java language 🤢

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -211,10 +218,20 @@ class ReplicaFetcherThread(name: String,
-    val fetchResponse = clientResponse.responseBody.asInstanceOf[FetchResponse]
-    fetchResponse.responseData.asScala.toSeq.map { case (key, value) =>
-      key -> new PartitionData(value)
+    log.info("ReplicaFetcherThread#fetch({})", fetchRequest.underlying)
```
Is this needed? If so, you can do INFO() w/o introducing log. Ditto in the info logging below.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -258,9 +276,14 @@ class ReplicaFetcherThread(name: String,
+    val fetchData = builder.build()
+    val requestBuilder = JFetchRequest.Builder.forReplica(fetchRequestVersion, replicaId,
+        maxWait, minBytes, fetchData.toSend()).setMaxBytes(maxBytes)
+    if (fetchMetadataSupported) {
```
Is this check needed given that we set fetchRequestVersion in the builder?

cmccabe replied:
It's a convenience to avoid putting references to specific versions into the body of the code.  Happy to replace it with `brokerConfig.interBrokerProtocolVersion >= KAFKA_1_1_IV0` if that makes sense?

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchResponse.java:
```diff
@@ -148,9 +152,19 @@
+    private static final Schema FETCH_RESPONSE_V7 = new Schema(
+        THROTTLE_TIME_MS,
+        ERROR_CODE,
+        new Field(FetchMetadata.TYPE, INT8, "The fetch type."),
```
Do we need the type in the response since it should always match that in the request?

cmccabe replied:
This is one of the changes from what was in the KIP.

The response type doesn't always match now.  A FULL request will get an INCREMENTAL response if an incremental fetch session was prepared.

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,319 @@
+            String fullRequestReason = null;
+            if (sessionParts == null) {
+                fullRequestReason = "Initializing new session";
+            } else if (prevMetadata.type() != INCREMENTAL) {
```
Hmm, when is prevMetadata ever set to INCREMENTAL? From the code, it seems that prevMetadata is only set to response.metadata(), which always starts with FULL.

cmccabe replied:
The server responds with INCREMENTAL when an incremental fetch request is possible.

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,319 @@
+                log.debug("Session handler for node {} received FetchResponse {} with {} partition(s).",
+                    node, response.metadata(), response.responseData().size());
+                if (prevMetadata.type() == FULL) {
+                    if (verifyFullFetchResponseParts(response)) {
```
Should we test for !verifyFullFetchResponseParts() here and the one in line 289?

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,319 @@
+                        prevMetadata = new FetchMetadata(FULL, prevMetadata.sessionId(), INVALID_EPOCH);
+                        return false;
+                    }
+                } else {
```
FetchType could also be SESSIONLESS. Should we check that?

cmccabe replied:
SESSIONLESS should be handled the same way as FULL.  Let me fix this.

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,319 @@
+                sessionParts = next;
+                next = null;
+                Map<TopicPartition, PartitionData> toSend =
+                    Collections.unmodifiableMap(new LinkedHashMap<>(sessionParts));
```
It seems that we can get here if FetchType is SESSIONLESS. In this case, it seems that we want to use the ordering of partitions in next to achieve fairness when there is more data to give than the max fetch response size?

cmccabe replied:
The ordering should be maintained, since `FetchSessionHandler#Builder#next` is a `LinkedHashMap`.  I guess there should be a comment about this in the code, so that it's documented.

junrao commented on core/src/main/scala/kafka/api/ApiVersion.scala:
```diff
@@ -184,3 +187,8 @@ case object KAFKA_1_0_IV0 extends ApiVersion {
+case object KAFKA_1_1_IV0 extends ApiVersion {
+  val version: String = "1.0-IV0"
+  val messageFormatVersion: Byte = RecordBatch.MAGIC_VALUE_V2
+  val id: Int = 13
```
13 => 14

junrao commented on clients/src/test/resources/log4j.properties:
```diff
@@ -12,10 +12,10 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
-log4j.rootLogger=OFF, stdout
+log4j.rootLogger=TRACE, stdout
```
This probably needs to be reverted?

junrao reviewed (commented):
@cmccabe : Thanks for the patch. Made a pass of non-testing files. Looks good to me overall. A few comments below.

cmccabe commented:
Rebased on changes to `KafkaApis.scala`

cmccabe commented:
I hit some bugs that were triggered by changing log4j.properties to use TRACE logging.  Filed fixes for them: #4450, #4449.  I also reverted the log4j.properties change in this patch, so hopefully we can get a clean junit test run soon :)

cmccabe committed:
KAFKA-6254: Incremental fetch requests

Implement incremental fetch requests as described by KIP-227.

cmccabe commented:
Rebased on trunk

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -211,10 +218,20 @@ class ReplicaFetcherThread(name: String,
-    val fetchResponse = clientResponse.responseBody.asInstanceOf[FetchResponse]
-    fetchResponse.responseData.asScala.toSeq.map { case (key, value) =>
-      key -> new PartitionData(value)
+    debug("ReplicaFetcherThread#fetch({})" + fetchRequest.underlying)
```
Should this be trace?

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -211,10 +218,20 @@ class ReplicaFetcherThread(name: String,
+    try {
+      val clientResponse = leaderEndpoint.sendRequest(fetchRequest.underlying)
+      val fetchResponse = clientResponse.responseBody.asInstanceOf[FetchResponse]
+      debug("ReplicaFetcherThread#fetch(" + fetchRequest.underlying +
```
Should this be trace?

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -65,17 +67,22 @@ class ReplicaFetcherThread(name: String,
   private val shouldSendLeaderEpochRequest: Boolean = brokerConfig.interBrokerProtocolVersion >= KAFKA_0_11_0_IV2
+  private val fetchSessionHandler = new FetchSessionHandler(logContext, sourceBroker.id)
+
+  val log = logContext.logger(ReplicaFetcherThread.getClass)
```
log doesn't seem to be used.

junrao commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -365,10 +388,13 @@ class ReplicaFetcherThread(name: String,
 
-  private[server] class FetchRequest(val underlying: JFetchRequest.Builder) extends AbstractFetcherThread.FetchRequest {
-    def isEmpty: Boolean = underlying.fetchData.isEmpty
+  private[server] class FetchRequest(val toSend: util.Map[TopicPartition, JFetchRequest.PartitionData],
```
Do we need to store toSend here?

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchMetadata.java:
```diff
@@ -0,0 +1,153 @@
+     * The FetchMetadata that is implicitly used for handling older FetchRequests that
+     * don't include fetch metadata.
+     */
+    public static final FetchMetadata LEGACY = new FetchMetadata(INVALID_SESSION_ID, FINAL_EPOCH);
```
From the KIP wiki, it seems that legacy request should use 0 as the epoch, not -1?

cmccabe replied:
Yeah.  I posted a correction about this.  The correct way is now id = 0, epoch = -1 (previously it was documented as id = -1, epoch = 0)

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchMetadata.java:
```diff
@@ -0,0 +1,153 @@
+
+    /**
+     * An invalid epoch.  When used in a fetch request, indicates that the client
+     * wants to close any existing session, and not create a new one.
```
Hmm, is this described in the KIP wiki? When will the client use a negative epoch?

cmccabe replied:
It's discussed at https://cwiki.apache.org/confluence/display/KAFKA/KIP-227%3A+Introduce+Incremental+FetchRequests+to+Increase+Partition+Scalability#KIP-227:IntroduceIncrementalFetchRequeststoIncreasePartitionScalability-FetchRequestMetadatameaning .  I will post a small correction to the mailing list since two numbers were swapped in the table earlier

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,342 @@
+        /**
+         * Mark that we want to remove a partition from the incremental fetch session.
+         */
+        public void remove(TopicPartition topicPartition) {
```
Is removed() being used?

cmccabe replied:
No, this isn't used. I'll take it out.

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,342 @@
+//                    // once we hit one of them, we know there are no more unseen entries to look
+//                    // at in next.
+//                    break;
+//                }
```
Is the commented out code still needed?

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,342 @@
+//                    break;
+//                }
+                if (!sessionParts.containsKey(topicPartition)) {
+                sessionParts.put(topicPartition, nextData);
```
indentation

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -539,41 +557,38 @@ class KafkaApis(val requestChannel: RequestChannel,
           response
         }
 
+        logger.trace("Sending Fetch response with parts.size={}, metadata={}",
```
This can just be TRACE(...).

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,690 @@
+    }
+    val responseSessionId = cache.maybeCreateSession(time.milliseconds(), isFromFollower,
+        updates.size(), createNewSession)
+    logger.trace("FullFetchContext generated updates.size={}, sessionId={}",
```
To be consistent, we want this class to extend Logging and use TRACE(...) for logging.

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,690 @@
+          logger.debug("Closed fetch session {}.", reqMetadata.id())
+        }
+      }
+      if (reqMetadata.epoch() == FINAL_EPOCH) {
```
Hmm, the KIP wiki doesn't document FINAL_EPOCH and and only says legacy requests are designated by a -1 session id.

hachikuji commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -139,9 +148,36 @@
+        new Field(MAX_BYTES_KEY_NAME, INT32, "Maximum bytes to accumulate in the response. Note that this is not an absolute maximum, " +
+            "if the first message in the first non-empty partition of the fetch is larger than this " +
+            "value, the message will still be returned to ensure that progress can be made."),
+        new Field(ISOLATION_LEVEL_KEY_NAME, INT8, "This setting controls the visibility of transactional records. Using READ_UNCOMMITTED " +
```
nit: These are the same descriptions as above. How about creating a static `Field` instances or at least extracting the message.

cmccabe replied:
max_bytes does have different doc strings in different message versions, though.  I started looking at adding more constants for this, but it got a bit messy-- maybe a good follow on change?

hachikuji commented on clients/src/main/java/org/apache/kafka/common/errors/InvalidFetchSessionEpochException.java:
```diff
@@ -0,0 +1,29 @@
+
+package org.apache.kafka.common.errors;
+
+public class InvalidFetchSessionEpochException extends ApiException {
```
Should this be retriable? Same question for `FetchSessionIdNotFoundException`.

cmccabe replied:
OK, let's make it retriable

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+        public FetchRequestData build() {
+            if (nextMetadata.isFull()) {
+                log.debug("Built full fetch {} for node {} with {} partition(s).",
+                    nextMetadata, node, next.size());
```
Perhaps we may as well list out all the partitions?

cmccabe replied:
As in the comment above, though, there are going to be a huge number of them.  So it's not really practical (except maybe on TRACE level)

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+            }
+            log.debug("Built incremental fetch {} for node {}. numAdded = {}, numRemoved = {}, " +
+                "numAltered = {}, sessionSize = {}", nextMetadata, node, numAdded, numRemoved,
+                numAltered, sessionParts.size());
```
Same as above. It will probably be particularly useful for incremental fetches to have the partitions explicitly in the log message.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+        /**
+         * All of the partitions which exist in the fetch request session.
+         */
+        private final Map<TopicPartition, PartitionData> sessionParts;
```
nit: can we spell out `partitions`?

cmccabe replied:
OK

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+    /**
+     * The maximum number of examples of missing partitions from a response to log.
+     */
+    private final static int MAX_MISSING_TO_LOG = 5;
```
I'm not sure this is a good idea. If we're unlucky, the partition we're interested in may not be listed. Since this is an exceptional case anyway, I would suggest using the more verbose message.

cmccabe replied:
OK.  If there is an error, we can log all partitions, to make it easier to debug.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+     * @param response  The response.
+     * @return          True if the incremental fetch response partitions are valid.
+     */
+    private boolean verifyIncrementalFetchResponseParts(FetchResponse response) {
```
nit: As before, are we saving that much by not spelling out `partitions`?

cmccabe replied:
It will now log the partitions if there is a mistake

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+        return true;
+    }
+
+    public boolean handleResponse(FetchResponse response) {
```
Can you document what the return value is indicating?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+                    "id {} and {} partition(s).", node, nextMetadata,
+                response.sessionId(), response.responseData().size());
+            if (nextMetadata.isFull()) {
+                if (!verifyFullFetchResponseParts(response)) {
```
I am wondering why this is not an assertion. Would the broker ever be expected to return only a subset of the partitions in a full fetch request? To be honest, I think it would be fine to skip these checks and just assume the broker gives us the right thing.

cmccabe replied:
We discussed this offline-- conclusion: let's think about this later.  For now, we should keep the existing behavior

hachikuji commented on core/src/main/scala/kafka/server/ReplicaFetcherThread.scala:
```diff
@@ -211,10 +218,20 @@ class ReplicaFetcherThread(name: String,
+    try {
+      val clientResponse = leaderEndpoint.sendRequest(fetchRequest.underlying)
+      val fetchResponse = clientResponse.responseBody.asInstanceOf[FetchResponse]
+      debug("ReplicaFetcherThread#fetch(" + fetchRequest.underlying +
```
nit: use string substitutions.

cmccabe replied:
Oops.  This log message should have been removed.  It duplicates some TRACE log messages in the network handling code that print out the request and response.  I'll get rid of it now

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+        }
+    }
+
+    public void handleError(Throwable t) {
```
Can you document when this should be used? From what I can see, it only has one use in the replica fetcher at the moment.

cmccabe replied:
> Can you document when this should be used?

It should be called whenever we couldn't communicate with the broker.  Basically it clears the fetch session state.

> From what I can see, it only has one use in the replica fetcher at the moment.

Good point-- it should have been called from `Fetcher.java` as well.  Fixed.

cmccabe replied:
I also added some JavaDoc

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+import static org.apache.kafka.common.requests.FetchMetadata.INVALID_SESSION_ID;
+
+/**
+ * Maintains the fetch session state for clients, including the fetch metadata and the cached partitions.
```
Some high-level comments would be useful here about how incremental fetching works (could be done in a follow-up if you like).

cmccabe replied:
Good point.  I added some high level comments and references to further reading

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+                        numAltered++;
+                    }
+                } else {
+                    // Set maxBytes to 0 to indicate that we no longer want to listen to this partition.
```
This comment is no longer relevant, right?

cmccabe replied:
Fixed

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -193,24 +198,30 @@ private boolean matchesRequestedPartitions(FetchRequest.Builder request, FetchRe
+                .isolationLevel(isolationLevel)
+                .setMaxBytes(this.maxBytes)
+                .metadata(data.metadata())
+                .toForget(data.toForget());
```
Maybe the name can be more explicit? For example, `forgetPartitions`?

cmccabe replied:
I wanted a name that indicated that we want to forget the partitions, but that it hasn't been done yet.  I'm open to suggestions, but `toForget` seemed nice and simple.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -193,24 +198,30 @@ private boolean matchesRequestedPartitions(FetchRequest.Builder request, FetchRe
+                .metadata(data.metadata())
+                .toForget(data.toForget());
+            log.debug("Sending {} fetch for partitions {} to broker {}", isolationLevel,
+                request.fetchData().keySet(), fetchTarget);
```
This message should refer to all the partitions in the fetch session, right?

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+                }
+            }
+        } else if (response.error() == Errors.FETCH_SESSION_ID_NOT_FOUND) {
+            log.info("Session handler for node {} received FETCH_SESSION_ID_NOT_FOUND error on {}.",
```
I am wondering if this can be lowered to `DEBUG` since it is handled internally.

cmccabe replied:
I think it makes sense to log since it's a pretty rare occurrence.  And if it does start happening a lot, that could indicate a problem.

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/consumer/internals/Fetcher.java:
```diff
@@ -193,24 +198,30 @@ private boolean matchesRequestedPartitions(FetchRequest.Builder request, FetchRe
-                                        "the requested partitions {}", response.responseData().keySet(),
-                                        request.fetchData().keySet());
+                            FetchSessionHandler handler = sessionHandlers.get(fetchTarget.id());
+                            if (handler == null) {
```
We lost the comment we had before, but it seemed useful. Maybe you can update it to be relevant to the new logic.

cmccabe replied:
Good point.  I will add a log message to FetchSessionHandler which will spell out this information

hachikuji commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,340 @@
+    }
+
+    public boolean handleResponse(FetchResponse response) {
+        if (response.error() == Errors.NONE) {
```
I was expecting to see some logic to remove a partition from the session following a `NOT_LEADER` error. Maybe I'm missing it somewhere?

junrao commented:
@cmccabe : Thanks for the patch. Only had time to review part of the patch. The following are my comments so far.

hachikuji reviewed (commented):
Did a quick pass over the client code and had a few questions/comments.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java:
```diff
@@ -0,0 +1,354 @@
+ * This collection requires previous and next indexes to be embedded into each
+ * element.  Using array indices rather than pointers saves space on large heaps
+ * where pointer compression is not in use.  It also reduces the amount of time
+ * the garbage collector has to spend chasing pointers.
```
I think the optimization of using array indices instead of pointers is a bit questionable without some benchmarks. Heaps larger than 32 GB are rarely (or never) used in Kafka. And having to go via the array has some cost as well.

cmccabe replied:
There are other benefits besides reducing the pointer size.  When you use array indices rather than pointers, the garbage collector needs to do less work chasing pointers.  See https://issues.apache.org/jira/secure/attachment/12701400/BlocksMap%20redesign.pdf .

Excerpt:

> According to an Oracle engineer, large heaps with reference dense objects in old gen with frequently mutating references is brutally hard on GC.When a reference in an old gen object is mutated, the object’s “card page” is marked as dirty. During young gen collection all references in dirty old gen card pages are used as roots for determining reachability of young gen objects.
>
> The [HDFS] block data-structure mutates by necessity, but it does so in a non-GC friendly manner. Report processing inserts a delimiter into the storage’s doubly linked list, moves reported blocks to the head of the storage’s list, then uses the delimiter to determine excess blocks for invalidation. The updating of so many references creates intense pressure on GC.
>
> One reason is young gen maintains a tenuring threshold equating to how conservatively it will promote young gen objects into old gen. The threshold drops relative to the rate of garbage creation and dirtying of old gen cards. The young collector may resort to prematurely promoting objects into old gen when it becomes overrun by spending too much time collecting. CMS is forced to cleanup when the old gen occupancy threshold is exceeded. The prematurely promoted objects lead to excessive fragmentation of old gen.
>
> We can reduce abusive GC behavior by reducing the mutation of references in old gen.
Unlike references, updating primitives (ints, longs, etc) does not mark an old gen page dirty.  It does not incur a penalty to young gen collection.
```

ijuma commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,690 @@
+    mustRespond
+  }
+
+  override def hashCode() = Objects.hash(topic, new Integer(partition))
```
The fact that we are allocating an array and potentially an `Integer` to compute the hashCode is suboptimal given that these elements are meant to be added to the `ImplicitLinkedHashSet`, which doesn't seem to cache hash codes.

ijuma commented on clients/src/main/java/org/apache/kafka/common/utils/ImplicitLinkedHashSet.java:
```diff
@@ -0,0 +1,354 @@
+    }
+
+    public ImplicitLinkedHashSet() {
+        this(10);
```
Note that `Hashtable` uses 11 (a prime number) as the default.

cmccabe replied:
I will change this to 5, so that we also get 11 as the default number of slots.

cmccabe commented:
retest this please

cmccabe commented:
Rebased on trunk

cmccabe commented:
I changed all the "parts" variables to "partitions".  I also changed it to log the full set of partitions in the response whenever there was a response verification problem.  When TRACE logging is enabled, it will now also log the full set of partitions that were sent, as well as the partitions that are in the session but not sent (aka "implied" partitions)

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/204/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/212/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/206/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/221/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/223/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/209/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/225/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/210/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/211/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/213/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/235/

cmccabe commented:
Test failure is `kafka.security.auth.SimpleAclAuthorizerTest.testHighConcurrencyModificationOfResourceAcls`, not related.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/249/

junrao commented on core/src/main/scala/kafka/server/KafkaApis.scala:
```diff
@@ -540,41 +558,38 @@ class KafkaApis(val requestChannel: RequestChannel,
           response
         }
 
+        logger.trace("Sending Fetch response with partitions.size={}, metadata={}",
```
To be consistent, in scala classes, we want to use TRACE(...).

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/257/

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchMetadata.java:
```diff
@@ -0,0 +1,153 @@
+     * @return          The next epoch.
+     */
+    public static int nextEpoch(int prevEpoch) {
+        if (prevEpoch < 0) {
```
When will prevEpoch ever go below 0? Could you document that?

cmccabe replied:
Only FINAL_EPOCH is less than 0.  I will add a comment.

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+    }
+    val responseSessionId = cache.maybeCreateSession(time.milliseconds(), isFromFollower,
+        updates.size(), createNewSession)
+    logger.trace("FullFetchContext generated updates.size={}, sessionId={}",
```
For consistency, for scala classes, we want them to extend Logging and use TRACE(...).

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchMetadata.java:
```diff
@@ -0,0 +1,153 @@
+        return (this.epoch == INITIAL_EPOCH) || (this.epoch == FINAL_EPOCH);
+    }
+
+    public int id() {
```
id can be a bit vague. Could we use sessionId?

cmccabe replied:
OK

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+  * @param creationMs   The time in milliseconds when this session was created.
+  * @param lastUsedMs   The last used time in milliseconds.  This should only be updated by
+  *                     FetchSessionCache#touch.
+  * @param epoch        The fetch session seequence number.
```
Typo seequence

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+  }
+
+  // Update the cached partition data based on the request.
+  def update(fetchData: FetchSession.REQ_MAP,
```
For ease of reading and consistency, could we explicitly define the return type of every method?

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+                        var epoch: Int) {
+  // This is used by the FetchSessionCache to store the last known size of this session.
+  // If this is -1, the Session is not in the cache.
+  var cachedSize = -1
```
Do we need cachedSize? Could we just use size() instead?

cmccabe replied:
We have to know the size the session was when it was inserted into the cache, in order to remove it from the cache.  (The most current size may be different)

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+
+      val oldSize = session.cachedSize
+      if (oldSize != -1) {
+        evictableByPrivileged.remove(session.evictableKey)
```
Could we call session.evictableKey once and reuse the result to avoid unnecessary object creation?

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+      if (oldSize != -1) {
+        evictableByPrivileged.remove(session.evictableKey)
+        evictableByAll.remove(session.evictableKey)
+        numPartitions = numPartitions - oldSize
```
Hmm, do we need session.cachedSize? Could we just call session.size() first before updating the new partition map?

cmccabe replied:
It's also used in `FetchManager#newContext`.  It would be awkward to not have it since it would mean that there was no good way to remove an arbitrary session (unless you had extra information about the session that wasn't contained in the session itself)

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,716 @@
+  */
+class FetchSessionCache(val maxEntries: Int,
+                        val evictionMs: Long) extends Logging with KafkaMetricsGroup {
+  var numPartitions: Long = 0
```
It seems that all var/val in this class can be private?

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,717 @@
+    }
+    if (respData.error.code() != 0) {
+      // Partitions with errors are always included in the response.
+      // We also set the cached lastStableOffset to an invalid offset, -1.
```
The code updates highWatermark instead of lastStableOffset.

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,717 @@
+    * Update this CachedPartition with new request and response data.
+    *
+    * This function should be called while holding the appropriate CachedPartitionMap
+    * lock.
```
Hmm, the caller of this method doesn't seem to be synchronized on the CachedPartition object. Do we have a visibility issue across threads?

cmccabe replied:
Oops, the comment is wrong.  That should read "the appropriate session lock."  Fixed.

junrao commented on clients/src/main/java/org/apache/kafka/clients/FetchSessionHandler.java:
```diff
@@ -0,0 +1,451 @@
+                    }
+                } else {
+                    // Indicate that we no longer want to listen to this partition.
+                    next.remove(topicPartition);
```
Since topicPartition doesn't exist in next if we get here, there is no need to remove it.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/271/

junrao commented on clients/src/main/java/org/apache/kafka/common/requests/FetchRequest.java:
```diff
@@ -153,7 +189,9 @@
     private final int maxBytes;
     private final IsolationLevel isolationLevel;
-    private final LinkedHashMap<TopicPartition, PartitionData> fetchData;
+    private final Map<TopicPartition, PartitionData> fetchData;
```
Since the caller always passes in a LinkedHashMap, is there a reason to change this to Map?

cmccabe replied:
The rationale is that `FetchSessionHandler` wraps the `LinkedHashMap` in an `UnmodifiableMap`, so the type is technically `UnmodifiableMap` rather than `LinkedHashMap` at that point.  Also, there's things like using `Collections.emptyMap` in a unit test which you can't use if you have to have a `LinkedHashMap`

I will add a comment saying that iteration order is significant, though.

junrao commented on clients/src/test/java/org/apache/kafka/clients/FetchSessionHandlerTest.java:
```diff
@@ -0,0 +1,361 @@
+        TopicPartition baz0 = new TopicPartition("baz", 0);
+        TopicPartition baz1 = new TopicPartition("baz", 1);
+        assertEquals("",
+            Utils.join(FetchSessionHandler.findMissing(toSet(foo0), toSet(foo0)), ", "));
```
Could we just test on set equal instead of string equal?

junrao commented on clients/src/test/java/org/apache/kafka/clients/FetchSessionHandlerTest.java:
```diff
@@ -0,0 +1,361 @@
+            0, 123);
+        handler.handleResponse(resp2);
+        assertMapsEqual(reqMap(new ReqEntry("foo", 0, 0, 100, 200)),
+            data.toSend(), data.sessionPartitions());
```
Hmm, are we supposed to test data? Should we build a new request?

cmccabe replied:
In this case, it was intentional to skip building a new request.  I'll add a comment to make it clearer (also we don't need to test `data#toSend` again)

junrao reviewed (commented):
@cmccabe : Thanks for the updated patch. A few more comments below.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/288/

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/289/

cmccabe commented:
The failed test on the jdk7 run is `kafka.security.auth.SimpleAclAuthorizerTest > testHighConcurrencyModificationOfResourceAcls`.  Not related.

junrao commented on core/src/main/scala/kafka/server/FetchSession.scala:
```diff
@@ -0,0 +1,720 @@
+            iter.remove()
+          }
+        }
+        debug("Incremental fetch context with session id %d returning %s".format(
```
On the server side, we have moved to the s convention for building a string, instead of format.

cmccabe replied:
OK.  I will change it over to the 's' convention

junrao commented on core/src/main/scala/kafka/server/KafkaConfig.scala:
```diff
@@ -861,6 +873,10 @@ object KafkaConfig {
 
+    /** ********* Fetch Session Configuration **************/
+      .define(MaxIncrementalFetchSessionCacheSlots, INT, Defaults.MaxIncrementalFetchSessionCacheSlots, atLeast(0), MEDIUM, MaxIncrementalFetchSessionCacheSlotsDoc)
+      .define(MinIncrementalFetchSessionEvictionMs, LONG, Defaults.MinIncrementalFetchSessionEvictionMs, atLeast(0), MEDIUM, MinIncrementalFetchSessionEvictionMsDoc)
```
I thought we agreed in the KIP that this will be a constant and not configurable?

cmccabe replied:
Good catch, will make this a constant.

junrao reviewed (commented):
@cmccabe : Thanks for the patch. Looks good to me. Just a couple of minor comments. 

@ijuma and @hachikuji : Do you want to take another look?

cmccabe commented:
retest this please

junrao commented:
@cmccabe : Thanks for update. The latest code LGTM. Do you have any performance results? It would be useful to see (1) the consumption improvement when there are idle topics, (2) no degradation when caching is disabled.

afalko commented:
@junrao I've been testing @cmccabe's patches. One thing that was important to us was the consumption latency that we define as (time it takes to consume a series of 100 small messages + time it takes to commit an offset). With kafka 1.0 and trunk, we'd see that latency exceed our SLA of 50 ms after 40-46k 3x replicated partitions. With Colin's patch, at fa01cf98 (before rebase) we were able to get to 68k 3x replicated partitions with latency of 35 ms. Generally, the offset commit latency is far higher than the consume message latency, 33 ms for the latter case and 46 ms for the former case.  I couldn't push more than 68k replicated partitions due to https://issues.apache.org/jira/projects/KAFKA/issues/KAFKA-6469?filter=allopenissues 

Let me know if you'd like me to get more results and share additional metrics. We're very excited about this patch!

junrao commented:
@afalko : Thanks for sharing the results. Very helpful. Just to clarify, are you saying the offset commit latency is 33ms without this patch and 46ms with the patch?

afalko commented:
Sorry @junrao I meant the other way around:

Without patch @46k: consume -  4 ms; commit - 46 ms
With patch @68k: consume - 2 ms; commit - 33 ms
With patch @46k: consume - 2 ms; commit - 21 ms

junrao commented:
@afalko : That's interesting. This patch doesn't really optimize the offset commit protocol. So, I am wondering why there is an improvement on offset commit.

cmccabe commented:
The test failure on JDK9 seems to be related to some ZK issues going on when running `kafka.api.PlaintextConsumerTest.testLowMaxFetchSizeForRequestAndPartition`.

from the logs:
```
[2018-02-03 07:33:00,012] WARN fsync-ing the write ahead log in SyncThread:0 took 1053ms which will adversely effect operation latency. See the ZooKeeper troubleshooting guide (org.apache.zookeeper.server.persistence.FileTxnLog:342)
...
[2018-02-03 07:33:37,276] WARN Client session timed out, have not heard from server in 4002ms for sessionid 0x101b1f2ea4e0000 (org.apache.zookeeper.ClientCnxn:1111)
```

The exception is:
```
org.apache.zookeeper.KeeperException$NodeExistsException: KeeperErrorCode = NodeExists for /config/topics/topic3
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:122)
	at org.apache.zookeeper.KeeperException.create(KeeperException.java:54)
	at kafka.zookeeper.AsyncResponse.maybeThrow(ZooKeeperClient.scala:450)
	at kafka.zk.KafkaZkClient.createRecursive(KafkaZkClient.scala:1388)
	at kafka.zk.KafkaZkClient.create$1(KafkaZkClient.scala:251)
	at kafka.zk.KafkaZkClient.setOrCreateEntityConfigs(KafkaZkClient.scala:258)
	at kafka.zk.AdminZkClient.createOrUpdateTopicPartitionAssignmentPathInZK(AdminZkClient.scala:99)
	at kafka.zk.AdminZkClient.createTopic(AdminZkClient.scala:56)
	at kafka.utils.TestUtils$.createTopic(TestUtils.scala:294)
	at kafka.integration.KafkaServerTestHarness.createTopic(KafkaServerTestHarness.scala:123)
	at kafka.api.PlaintextConsumerTest.$anonfun$testLowMaxFetchSizeForRequestAndPartition$1(PlaintextConsumerTest.scala:791)
```

I don't think this is related to the patch at all.  I will re-run the tests to see if we can get a clean run this time.

cmccabe commented:
retest this please

cmccabe commented:
@afalko: thanks again for your great work testing this.

@junrao  wrote: 
> That's interesting. This patch doesn't really optimize the offset commit protocol. So, I am wondering why there is an improvement on offset commit.

Yeah, that is interesting.  After all, we are handling the same number of partitions on the broker, but we are just not serializing them into every RPC like we did before.  So I would expect the offset commit time improvement to come from better behaved garbage collection performance or better network utilization.  Probably network utilization, since the patch doesn't make too many special efforts to optimize GC (although I made one here and there-- for example using iterators instead of copying a map in one place.)

hachikuji commented:
Offset commits depend on replication, so any improvement to fetch overhead could reduce offset commit latency. If the result is actually meaningful, I would expect to see a similar improvement for produce latency.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/kafka-pr-test-coverage/320/

junrao commented:
@hachikuji : Great point. That makes sense.

afalko commented:
Thanks @hachikuji, @junrao, @cmccabe . Fresh off open source presses, I've been able to open source the test I wrote that was measuring the results that I mentioned: https://github.com/salesforce/kafka-partition-availability-benchmark

I plan to expand it to have another mode where it is producing continuously without resetting offset. That'll be able to measure the produce latency.

junrao commented:
@afalko : Thanks. You may want to link that to the jira so that other people know how your tests were done.
